src/main/scala/com/fulcrumgenomics/bam/ErrorRateByReadPosition.scala

```diff
@@ -25,266 +25,234 @@
-
-    if (ref.exists(r => in.getFileHeader.getSortOrder != SortOrder.coordinate)) {
-      fail("Input BAM must be coordinate sorted when the reference is given.")
+    val metrics     = computeMetrics
+    val prefix      = output.getOrElse(PathUtil.removeExtension(input))
+    val metricsPath = PathUtil.pathTo(prefix + ErrorRateByReadPositionMetric.FileExtension)
+    val plotPath    = PathUtil.pathTo(prefix + ErrorRateByReadPositionMetric.PlotExtension)
+    Metric.write(metrics, metricsPath)
+
+    // And try plotting
+    if (metrics.nonEmpty) {
+      val name = PathUtil.basename(input, trimExt=true).toString
+      Rscript.execIfAvailable(ScriptPath, metricsPath.toString, plotPath.toString, name) match {
+        case Failure(e) => logger.warning(s"Generation of PDF plots failed: ${e.getMessage}")
+        case _ => Unit
+      }
     }
+  }
 
-    iterator.filterNot { rec =>
-      rec.getReadUnmappedFlag || rec.isSecondaryOrSupplementary ||
-        (!includeDuplicates && rec.getDuplicateReadFlag) || rec.getReadFailsVendorQualityCheckFlag ||
-        (rec.getMappingQuality < minMappingQuality)
-    }
-    .foreach { rec =>
-      counter.accept(rec)
-      progress.record(rec)
-    }
+  /** Computes the metrics from the BAM file. */
+  private[bam] def computeMetrics: Seq[ErrorRateByReadPositionMetric] = {
+        val progress = new ProgressLogger(logger, verb="Processed", noun="loci", unit=50000)
+    val in = SamReaderFactory.make.open(input)
+    val ilist = this.intervals.map(p => IntervalList.fromFile(p.toFile))
 
-    val path = PathUtil.pathTo(output.getOrElse(PathUtil.removeExtension(input)) + ErrorRateByReadPositionMetric.FileExtension)
-    Metric.write(metrics=counter.metrics, path=path)
+    val refWalker = new ReferenceSequenceFileWalker(this.ref.toFile)
+    val locusIterator = buildSamLocusIterator(in, ilist)
+    val variantMask = buildVariantMask(variants, ilist, refWalker.getSequenceDictionary)
 
-    in.safelyClose()
-  }
-}
+    val counters = List.tabulate(3)(_ => mutable.Map[Int, ObsCounter]())
 
-object ErrorRateByReadPositionMetric {
-  val FileExtension = ".error_rate_by_read_position.txt"
-}
-
-/** Gives the error rate and total number of bases observed for each position within the read.
-  *
-  * @param read_number the read number (ex. 0 for fragments, 1 for read one of a pair, 2 for read two of a pair)
-  * @param position the read position (1-based).
-  * @param error_rate the error rate at the given position n the read
-  * @param bases_total the total number of bases observed at this position.
-  */
-case class ErrorRateByReadPositionMetric
-( read_number: Int,
-  position: Int,
-  error_rate: Double,
-  bases_total: Long
-) extends Metric
+    locusIterator.filterNot(l => variantMask.isVariant(l.getSequenceIndex, l.getPosition)).foreach { locus =>
+      val refBase = SequenceUtil.upperCase(refWalker.get(locus.getSequenceIndex).getBases()(locus.getPosition - 1))
```

Comment:
refWalker is so simple, would probably be easier to just write our own in Scala that provides some nicer APIs.